From 85029a43e536d324d2cf96d57ba9bfe3fb3735fa Mon Sep 17 00:00:00 2001
From: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Date: Wed, 7 Jun 2023 22:14:12 +0530
Subject: [PATCH 2/2] wifi: ath12k: fix link specific tx mgmt pending free

When the link get removed, we wrongly remove the tx mgmt pending belong to
the vif. we have to remove only the link specific tx mgmt pending. So
added the link type in the mac tx mgmt pending free and pass the correct
link type from link remove procedure.

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 11 ++++++++++-
 drivers/net/wireless/ath/ath12k/mac.h |  2 ++
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 491d672ef8fa..cea85c9b02ba 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -8493,6 +8493,13 @@ int ath12k_mac_tx_mgmt_pending_free(int buf_id, void *skb, void *ctx)
 			goto out;
 	}
 
+	/* If link_id valid, then free the match link_id alone */
+	if (u8_get_bits(arg->type, ATH12K_MAC_TX_MGMT_FREE_TYPE_LINK)) {
+		if ((skb_cb->link_id != arg->link_id) &&
+		    (skb_cb->link_id != IEEE80211_MLD_MAX_NUM_LINKS))
+			goto out;
+	}
+
 	ath12k_mac_tx_mgmt_free(arg->ar, buf_id);
 
 out:
@@ -10571,8 +10578,10 @@ err_vdev_del:
 
 	arg.ar = ar;
 	arg.vif = vif;
+	arg.link_id = arvif->link_id;
 	arg.type = u8_encode_bits(true, ATH12K_MAC_TX_MGMT_FREE_TYPE_PDEV) |
-		   u8_encode_bits(true, ATH12K_MAC_TX_MGMT_FREE_TYPE_VIF);
+		   u8_encode_bits(true, ATH12K_MAC_TX_MGMT_FREE_TYPE_VIF) |
+		   u8_encode_bits(true, ATH12K_MAC_TX_MGMT_FREE_TYPE_LINK);
 
 	idr_for_each(&ar->txmgmt_idr,
 		     ath12k_mac_tx_mgmt_pending_free, &arg);
diff --git a/drivers/net/wireless/ath/ath12k/mac.h b/drivers/net/wireless/ath/ath12k/mac.h
index 4339bf8154e8..f96d82fe3e65 100644
--- a/drivers/net/wireless/ath/ath12k/mac.h
+++ b/drivers/net/wireless/ath/ath12k/mac.h
@@ -32,11 +32,13 @@ struct ath12k_mac_num_chanctxs_arg {
 
 #define ATH12K_MAC_TX_MGMT_FREE_TYPE_PDEV	BIT(0)
 #define ATH12K_MAC_TX_MGMT_FREE_TYPE_VIF	BIT(1)
+#define ATH12K_MAC_TX_MGMT_FREE_TYPE_LINK	BIT(2)
 
 struct ath12k_mac_tx_mgmt_free_arg {
 	u8 type;
 	struct ath12k *ar;
 	struct ieee80211_vif *vif;
+	u8 link_id;
 };
 
 /* Default link after the IEEE802.11 defined Max link id limit
-- 
2.17.1

